### ISSUE TITLE
[Proposal] Folder structure following the Atomic Design name convention

### ISSUE THREAD

**Issue by cristianoliveira** [sumup/circuit-ui#62]
Since we are at the very beginning of the library I think it is the best moment to propose it and try to build a ground for a maintainable code. :)  

I noticed that we are currently adding all components under folder `components` which isn't a problem, but since we are following the Atomic Design principles for this lib. How about separate the components, not for implementation details but instead structure around the atomic design component kind. Like `atoms`, `molecules` and `organisms`?

> ** Impact of a Name **
> 
> There is nothing new with modular design and development. Hence, what is the impact of certain terms used as in “atoms”, “molecules”, and “organisms”? Design and development terms like “modules”, “components”, and “elements” could easily be adopted as standard. However, the discussion on atomic design requires the choice of chemical terminology to imply a hierarchy which assists those with basic chemistry knowledge to understand on atomic design. Other terminologies lack that ability to help UI designers, developers and end users appreciate and understand the full effect as clearly.
> 
> reference: http://ubie.io/atomic-design/

```
src/components/
  atoms/
    ▾ Input/
        index.js
        Input.js
        Input.spec.js
    ▾ Button/
        Button.js
        Button.spec.js
        index.js
    ▾ Icon/
        Icon.js
        Icon.spec.js
        index.js
  molecules/
    ▾ SearchInput/
        index.js
        SearchInput.js (Icon + Input)
  organisms/
     ▾ HorizontalNavigation/
        index.js
        HorizontalNavigation.js
```

That makes easier to think about the components in terms of usage and concerns. Plus a new joiner will easily understand the idea behind the components design/principles. 

Also in the future, we could create small bundles only with atoms, molecules or organisms them who uses the library could import only the set of components that he will use instead of the whole lib.

Some references behind this structure usage:
  - https://cheesecakelabs.com/br/blog/atomic-design-react/
  - https://medium.com/@yejodido/atomic-components-managing-dynamic-react-components-using-atomic-design-part-1-5f07451f261f
  - https://github.com/Rulox/react-atomic-structure
  - https://github.com/danilowoz/react-atomic-design
  - http://ubie.io/atomic-design/

**Comment by connor-baer** [comment]
I like this idea. We kinda do this in the website repository already.

**Comment by mlent** [comment]
I like it conceptually, but I wonder if new joiners will be even more confused trying to figure out what's an atom and what's a molecule? For instance, Liam said the Card is a molecule whereas I would've thought Atom :/

To me, just having an alphabetical list is the most helpful - also when we display the components, I would imagine that would be the most searchable/scannable.

**Comment by cristianoliveira** [comment]
> I like it conceptually, but I wonder if new joiners will be even more confused trying to figure out what's an atom and what's a molecule? For instance, Liam said the Card is a molecule whereas I would've thought Atom :/

That exactly the kind of question this structure should raise. So the new joiners can look into the concept and understand what is all about. As long we keep it consistent the differentiation between then shouldn't be that hard. 

> To me, just having an alphabetical list is the most helpful - also when we display the components, I would imagine that would be the most searchable/scannable.

If you are talking about the `storybook` displaying. Quite sure it has a configuration to display in the order you want.

**Comment by mlent** [comment]
Nah, not talking about storybook but in the folder structure. E.g. I'm looking for a card, and I think it's one thing but the developer who made it thought it's another.

Like, I also agree that this separation is nice like what we have on the marketing website. Though I'm not sure whether directories for atoms/molecules/organisms is as sensible as a different kind of distinction, like "elements" vs "components".

@felixjung WDYT?

**Comment by felixjung** [comment]
Hmmmm. I definitely like the spirit. However, unless we look to make our entire documentation and the way we "advertize" this library about atomic design, I don't think I see a real benefit.

From the user perspective:
- The library, for now, is imported as a ESM. You can import whatever part you want, down to the component and would tree-shake the rest. The way all bundlers are going is that tree-shaking will have real benefits and come out of the box.
- Right now I don't think the library will be large enough to warrant splitting it up into parts.

From the developer perspective:
- With fuzzy search I don't care about where the components are during development.
- Without fuzzy search, I might actually find it annoying to have to figure out where to find a component.
- Import paths will get longer and you need to think about where stuff is.
- Creating a new component via a CLI might be a bit weird. We'd have to write a CLI that knows where to put stuff (probably you could achieve that via a flag, `sumup create component --atom`).
- The bigger libraries (just took another look at material, polaris, and atlaskit) all have a flat project structure with respect to their components.

None of these points are real deal breakers, I think. If we want to emphasize the whole atomic design thing, it makes sense to group them and align the docs. Then it would also make sense to have different library entries (i.e., `import * as atoms from 'circuit-ui/atoms';`). But to me it seems more like a marketing decision at this point. Might change once the folder gets crowded.

**Comment by mlent** [comment]
I would still imagine that we export everything top level and not expose this "atoms" stuff to end users. Why would someone _using_ our UI library need to care about its philosophy? Usage should be unopinionated (why we offer a css version, why we didn't want to force a css-in-js solution for end consumers, etc.). I also would not want to organize them this way in our docs - it just makes it harder for outsiders to find things, IMO.

So let's table this and consider it once we're wrapping up and see how we feel about things then. As I would definitely not do something like `import { Button } from 'circuit-ui/atoms'` even if we did have an extra directory level, I don't think there's any reason to do this before we have the context of all the components and can decide if we need an additional layer of grouping.

**Comment by cristianoliveira** [comment]
I agree with deciding it after when we get more use cases. For now, is more a thinking in advance that probably falls down in YAGN thing. I just saw this pattern being used in those links I referenced in the first post and kind of liked the approach. :)

>Why would someone using our UI library need to care about its philosophy?

Who uses the library actually don't need to but the internal developers would be nice to know since it helps the communication between designers and devs. If not, those name conventions wouldn't be needed in the first place but look like they are, at least here: https://github.com/sumup/circuit-ui/projects :D 

The package naming is just a name convention. For instance, we could use as `modules`, `components` and `elements`(etc) would have the same effect: grouping the components for its purpose. When choosing those name, IMO, make sense to follow the "philosophy". :)

> I would still imagine that we export everything top level and not expose this "atoms" stuff to end users. Why would someone using our UI library need to care about its philosophy? Usage should be unopinionated (why we offer a css version, why we didn't want to force a css-in-js solution for end consumers, etc.). I also would not want to organize them this way in our docs - it just makes it harder for outsiders to find things, IMO.

You won't expose atoms, but probably molecules/organisms. For instance, when creating some small component to embed in a third party site you won't need the whole library but just part of it like some badges(molecules), typography and logo. (probably other things)

>So let's table this and consider it once we're wrapping up and see how we feel about things then.

Fair enough. :)
